templates/web/src/sdk.ts.twig

```diff
@@ -137,71 +145,122 @@ class {{ spec.title | caseUcfirst }} {
     private realtime: Realtime = {
         socket: undefined,
         timeout: undefined,
-        channels: {},
+        url: '',
+        channels: new Set(),
+        subscriptions: new Map(),
+        subscriptionsCounter: 0,
+        reconnect: true,
+        reconnectAttempts: 0,
         lastMessage: undefined,
+        connect: () => {
+            clearTimeout(this.realtime.timeout);
+            this.realtime.timeout = window?.setTimeout(() => {
+                this.realtime.createSocket();
+            }, 50);
+        },
+        getTimeout: () => {
+            switch (true) {
+                case this.realtime.reconnectAttempts < 5:
+                    return 1000;
+                case this.realtime.reconnectAttempts < 15:
+                    return 5000;
+                case this.realtime.reconnectAttempts < 100:
+                    return 10_000;
+                default:
+                    return 60_000;
+            }
+        },
         createSocket: () => {
+            if (this.realtime.channels.size < 1) return;
             const channels = new URLSearchParams();
             channels.set('project', this.config.project);
-            for (const property in this.realtime.channels) {
-                channels.append('channels[]', property);
-            }
-            if (this.realtime.socket?.readyState === WebSocket.OPEN) {
-                this.realtime.socket.close();
-            }
+            this.realtime.channels.forEach(channel => {
+                channels.append('channels[]', channel);
+            });
 
-            this.realtime.socket = new WebSocket(this.config.endpointRealtime + '/realtime?' + channels.toString());
-            this.realtime.socket?.addEventListener('message', this.realtime.authenticate);
+            const url = this.config.endpointRealtime + '/realtime?' + channels.toString();
 
-            for (const channel in this.realtime.channels) {
-                this.realtime.channels[channel].forEach(callback => {
-                    this.realtime.socket?.addEventListener('message', callback);
+            if (url !== this.realtime.url || !this.realtime.socket || this.realtime.socket?.readyState > WebSocket.OPEN) {
+                if (this.realtime.socket && this.realtime.socket?.readyState < WebSocket.CLOSING) {
```

Comment:
Can we add some comments here to help understand these conditionals?  Kind of getting lost in here